pymemcache/client/retrying.py

```diff
@@ -0,0 +1,297 @@
+    """
+
+    argument_tuple = None
+
+    # Ensure the argument is a tuple, or raise an error if we can't convert it.
+    if argument is None:
+        return tuple()
+    elif isinstance(argument, tuple):
+        argument_tuple = argument
+    elif isinstance(argument, Iterable) and not isinstance(argument, (str, bytes, bytearray)):
+        argument_tuple = tuple(argument)
+    else:
+        raise ValueError("%s was not a tuple or Iterable." % argument_name)
+
+    # Check that all the elements are actually inherited from Exception. (Catchable)
+    if not all([issubclass(arg, Exception) for arg in argument_tuple]):
+        raise ValueError("%s contained elements other than Exceptions." % argument_name)
+
+class RetryingClient(object):
+    """
+    Client that allows retrying calls for the other clients.
+    """
+
+    def __init__(
+        self,
+        client,
+        attempts=2,
+        retry_delay=0,
+        retry_for=None,
+        do_not_retry_for=None
+    ):
+        self.client = client
+        self.attempts = attempts
+        self.retry_delay = retry_delay
+        self.retry_for = _ensure_tuple_argument("retry_for", retry_for)
+        self.do_not_retry_for = _ensure_tuple_argument("do_not_retry_for", do_not_retry_for)
+
+
+    # This is the real magic soup of the class, we catch anything that isn't strictly defined for
+    # ourselves and pass it on to whatever client we've been given.
+    def __getattr__(self, name, *args, **kwargs):
+        # For no retry:
+        # return self.client.__getattribute__(name, *args, **kwargs)
+        for attempt in range(self.attempts):
+            try:
+                return self.client.__getattribute__(name, *args, **kwargs)
+            except Exception as exc:
+
+                if attempt == self.attempts:
+                    # If we're at the limit, just raise the exception to caller.
+                    # Exception type doesn't matter.
+                    raise exc
+                elif isinstance(exc, self.retry_for) or not isinstance(exc, self.do_not_retry_for):
+                    # If the exception is either a valid for retry based on type,
+                    # or just not invalid, retry it.
+                    sleep(self.retry_delay)
+                    continue
+                else:
+                    # Base case, raise exception to caller.
+                    raise exc
+
+
+    # We implement these explicitly because they're "magic" functions and won't get passed on by
+    # __getattr__.
+
+    def __dir__(self):
+        return dir(self.client)
+
+    def __setitem__(self, key, value):
+        return self.set(key, value, noreply=True)
+
+    def __getitem__(self, key):
+        value = self.get(key)
+        if value is None:
+            raise KeyError
+        return value
+
+    def __delitem__(self, key):
+        return self.delete(key, noreply=True)
+
+
+    # "static" implementations.
```

Comment:
You can just delete the comments below this line and stick with the dynamic implementation.